precise/iaf_psc_exp_ps_lossless.cpp

```diff
@@ -0,0 +1,671 @@
+          }
+          else
+          {
+            S_.I_syn_in_ += ev_weight; // inh. spike input
+          }
+        }
+
+        // store state
+        V_.I_syn_ex_before_ = S_.I_syn_ex_;
+        V_.I_syn_in_before_ = S_.I_syn_in_;
+        V_.y2_before_ = S_.y2_;
+        last_offset = ev_offset;
+      } while ( B_.events_.get_next_spike(
+        T, true, ev_offset, ev_weight, end_of_refract ) );
+
+      // no events remaining, plain update step across remainder
+      // of interval
+      if ( last_offset > 0 ) // not at end of step, do remainder
+      {
+        const double spike_time_max = is_spike_( last_offset );
+        propagate_( last_offset );
+        if ( not numerics::is_nan( spike_time_max ) )
+        {
+          emit_spike_( origin, lag, V_.h_ms_ - last_offset, spike_time_max );
+        }
+      }
+    } // else
+
+    // Set new input current. The current change occurs at the
+    // end of the interval and thus must come AFTER the threshold-
+    // crossing approximation
+    S_.y0_ = B_.currents_.get_value( lag );
+
+    // log state data
+    B_.logger_.record_data( origin.get_steps() + lag );
+  } // for
+}
+
+// function handles exact spike times
+void
+nest::iaf_psc_exp_ps_lossless::handle( SpikeEvent& e )
+{
+  assert( e.get_delay() > 0 );
+
+  /* We need to compute the absolute time stamp of the delivery time
+     of the spike, since spikes might spend longer than min_delay_
+     in the queue.  The time is computed according to Time Memo, Rule 3.
+  */
+  const long Tdeliver = e.get_stamp().get_steps() + e.get_delay() - 1;
+
+  B_.events_.add_spike(
+    e.get_rel_delivery_steps(
+      nest::kernel().simulation_manager.get_slice_origin() ),
+    Tdeliver,
+    e.get_offset(),
+    e.get_weight() * e.get_multiplicity() );
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::handle( CurrentEvent& e )
+{
+  assert( e.get_delay() > 0 );
+
+  const double c = e.get_current();
+  const double w = e.get_weight();
+
+  // add weighted current; HEP 2002-10-04
+  B_.currents_.add_value(
+    e.get_rel_delivery_steps(
+      nest::kernel().simulation_manager.get_slice_origin() ),
+    w * c );
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::handle( DataLoggingRequest& e )
+{
+  B_.logger_.handle( e );
+}
+
+// auxiliary functions ---------------------------------------------
+
+void
+nest::iaf_psc_exp_ps_lossless::propagate_( const double dt )
+{
+  const double expm1_tau_ex = numerics::expm1( -dt / P_.tau_ex_ );
+  const double expm1_tau_in = numerics::expm1( -dt / P_.tau_in_ );
+
+  if ( not S_.is_refractory_ )
+  {
+    const double expm1_tau_m = numerics::expm1( -dt / P_.tau_m_ );
+
+    const double P20 = -P_.tau_m_ / P_.c_m_ * expm1_tau_m;
+    const double P21_ex = -P_.tau_m_ * P_.tau_ex_ / ( P_.tau_m_ - P_.tau_ex_ )
+      / P_.c_m_ * ( expm1_tau_ex - expm1_tau_m );
+    const double P21_in = -P_.tau_m_ * P_.tau_in_ / ( P_.tau_m_ - P_.tau_in_ )
+      / P_.c_m_ * ( expm1_tau_in - expm1_tau_m );
+
```

Comment:
Same as before, I think you can multiply with `P_.c_m_` instead of dividing.

Also, I generally find it a bit confusing with the division in the middle with multiplications around, but I know this is how it would be written on paper (when you have the e^(...) ), so it is fine.